### DISCUSSION TITLE
Maplibre x MapBox API

### DISCUSSION THREAD

**Discussion by weleo** [maplibre/maplibre-gl-js#1226] [Q&A]
**maplibre-gl-js version**: `2.1.9`

I am trying to use Maplibre with the Mapbox API as describes [here](https://maplibre.org/maplibre-gl-js-docs/style-spec/sources/#vector),
but I am getting `URL scheme "mapbox" is not supported.`. 

1. Are the docs up-to-date?
2. To what extent do I need to change Mapbox style specifications so that they can be consumed by Maplibre?

**Comment by acalcutt** [comment]
I'm pretty sure support for "mapbox://" urls has been dropped from maplibre 2.x . I think the consensus is if you are using mapbox sources, you should be using the offical mapbox-gl-js to comply with their license.

You might be able to use something like this
```
sources: {
	mapbox: {
		type: 'vector',
		attribution: 'Mapbox',
		tiles: ['https://a.tiles.mapbox.com/v4/mapbox.mapbox-streets-v7/{z}/{x}/{y}.vector.pbf?access_token=your_key']
	}
}
```
example: https://stackblitz.com/edit/web-platform-dbcb9a

**Comment by HarelM** [comment]
Yup, mapbox scheme support was dropped.
Converting this into a discussion.
Note that there's already a discussion that was done before dropping the support to notify the community.

**Reply by itbeyond** [reply]
Hum - very long term mapbox-gl-js user and I am using a custom version from a fork of 1.11.0 using offline tiles via custom db driver. However I also provide access to two mapbox sources for Satellite and Streets. There should be no licensing issue as far as I can tell. I think this was the wrong call, where can I view this previous discussion?

**Reply by HarelM** [reply]
https://github.com/maplibre/maplibre-gl-js/discussions/290
https://github.com/maplibre/maplibre-gl-js/discussions/71

There are no plans to bring this back.
Any one in the community can write a simple plugin that uses `addProtocol` to bring this back.
The fact that no one had done so indicate this is not that important and that people mainly like to write about how unhappy they are about the new situation and not do something about it.
Also, as I stated many times before, if you are using Maobox data and services you should pay them, they are doing a great job! 😀

**Reply by itbeyond** [reply]
I have read these discussions and there are people in both camps and many asking for how to do this. However, I have retracted my wrong call statement above and I agree with a lot of the discussions and think the mapbox:// removal is worth while as you are right it complicates things. However I think there are plenty of people who want this functionality and suspect they are sticking with MapBox or MapLibre V1. This to me is a loss for the efforts being made here with this library.

I do pay MapBox for map views and other api usage as per their TOU's which is how the licensing works. Incidentally I can pay them for my map views without using mapbox-gl-js and I think it is just wrong to assume that anyone will not be paying MapBox for services if they go with MapLibre. For me it is just that the MapBox sources I provide are secondary to the vector tiles that I produce and are only rendered around 5% of the time for my user base thus the 2.0 model and the massive amount of work I would need to bring my offline tiles loaders into V2 is just not suitable.

I am not complaining I am happy to contribute to this project and even sponsor it if I can get this sorted, as a solid mapping engine library is of major importance to my business. addProtocol and on-going community support are the things that I really want. addProtocol is how my local DB tiles can be rendered and is how I can get away from the custom tile loaders I am running in my heavily modified mapbox gl js 1.11.0. 

I would be happy to create this plugin if I could work out how to get converted styles to actually run error free. Problem is at this stage that it is does not seem as simple as just write an addProtocol function. I cannot make hard coded complex styles from mapbox render in maplibre without issues (I may have something wrong but it seems far from simple).  Anyway I will keep working on it and once I can get the MapBox styles to work correctly in MapLibre I would be able to use this library and think about contributing via a plugin or similar, just got to get passed step 1 making my products work.

**Reply by HarelM** [reply]
When I removed the mapbox protocol support it was mainly two files, without a lot of code in them. These files can be found in version 1.x.
I personally do not use mapbox data at all so I don't have any incentive to write this plugin ATM.
Feel free to contact me privately through the osmus slack channel.

**Reply by acalcutt** [reply]
Is there a reason why a workaround like https://github.com/maplibre/maplibre-gl-js/discussions/1226#discussioncomment-2695633 wouldn't work for you? It's mostly updating the urls in your style to not have mapbox:// and including your key. I admit changing the style is a but of a pain, but if you only have a few styles that use mapbox it doesn't seem like a huge change. (see the example in that post)

**Reply by itbeyond** [reply]
I have now managed to make this work via a hard code of the styles. I also just want to clarify that when you do this and use MapLibre vs Mapbox the MapBox billing adjusts from a Map Load model to a Tile Load model as the map.load event is not fired from MapLibre which it is in MapBox even in V1, The V1 to V2 difference is that map.load is fired when a mapbox map is rendered V1 vs map.load is fired when the map object is created V2.  Moving to a tile load model I believe for most people will actually cost more for the MapBox services which is also fine and does reinforce my comment about paying for services, so I think it is wrong to insinuate that people will not be paying MapBox if they want to use MapBox styles in the MapLibre library.

I need to do a bit more playing and work with addProtocol to understand how a plugin could be created.

**Reply by wipfli** [reply]
The way I understood the message was the people using Mapbox should pay for their great services. I think I did not see someone saying that using MapLibre with their services will allow you to circumvent paying for tiles...

**Reply by wipfli** [reply]
But anyway, I am glad that you found a solution @itbeyond and it is great if you can switch to MapLibre GL JS!

**Reply by wipfli** [reply]
So welcome to the club :)

**Comment by rowanwins** [comment]
I have just published the first version of a small library with the aims of helping to transform styles for use in MaplibreGL
https://www.npmjs.com/package/maplibregl-mapbox-request-transformer

Feedback or collaborators welcome. 

````
import { isMapboxURL, transformMapboxUrl } from 'maplibregl-mapbox-url-transformer'

const mapboxKey = 'pk.123'

const transformRequest = (url, resourceType) => {
  if (isMapboxURL(url)) {
    return transformMapboxUrl(url, resourceType, mapboxKey)
  }
  
  // Do any other transforms you want
  return {url}
}

var map = new maplibregl.Map({
  container: 'map',
  center: [-122.420679, 37.772537],
  zoom: 13,
  style: 'mapbox://styles/mapbox/satellite-streets-v11',
  transformRequest
});
````

**Reply by ghost** [reply]
This looks very useful.

Please add it to https://github.com/maplibre/awesome-maplibre and make sure you mention buzzwords that people might search for.
I'd recommend adding https://github.com/rowanwins/maplibregl-mapbox-request-transformer instead of the npm address though.
